### ISSUE TITLE
no-shadowed-variable error with type assertions

### ISSUE THREAD

**Issue by jgornick** [palantir/tslint#3028]
### Bug Report

- 5.5.0:
- 2.3.4:
- CLI

#### TypeScript code being linted

```ts
import { Component, Inject } from '@angular/core';
import { FormGroup, FormBuilder, Validators, ValidatorFn, AbstractControl } from '@angular/forms';

@Component({
  // ...
})
export class FormComponent {
  public form: FormGroup;

  constructor(
    @Inject(FormBuilder) private formBuilder: FormBuilder
  ) {
    this.form = this.formBuilder.group({
      field: [
        'defaultValue'
        [
          Validators.required,
          <ValidatorFn> (control: AbstractControl) => {
            // ...
          }
        ]
      ]
    });
  }
```

with `tslint.json` configuration:

```json
{
  "extends": [
    /* @see https://github.com/palantir/tslint/blob/master/src/configs/recommended.ts */
    "tslint:recommended",
    /* @see https://github.com/buzinas/tslint-eslint-rules/blob/master/tslint_eslint_rules.json */
    "tslint-eslint-rules"
    /* @see https://github.com/Microsoft/tslint-microsoft-contrib/blob/master/recommended_ruleset.js */
    // "tslint-microsoft-contrib"
  ],
  "rulesDirectory": [
    /* @see https://github.com/mgechev/codelyzer */
    "node_modules/codelyzer"
    /* @see https://github.com/vrsource/vrsource-tslint-rules */
    // "node_modules/vrsource-tslint-rules/rules",
    /* @see https://github.com/Microsoft/tslint-microsoft-contrib */
    // "node_modules/tslint-microsoft-contrib"
  ],
  "rules": {
    // ESLint Overrides/Custom
    "brace-style": [
      true,
      "1tbs",
      {
        "allowSingleLine": false
      }
    ],
    "object-curly-spacing": [true, "always"],
    "space-in-parens": [true, "never"],
    "ter-indent": [
      true,
      2,
      {
        "SwitchCase": 1
      }
    ],

    // Angular (Codelyzer)
    "component-class-suffix": true,
    "component-selector": [true, "element", "pre", "kebab-case"],
    "directive-class-suffix": true,
    "directive-selector": [true, "attribute", "pre", "camelCase"],
    "invoke-injectable": true,
    "no-access-missing-member": true,
    "no-attribute-parameter-decorator": true,
    "no-forward-ref": true,
    "no-input-rename": true,
    "no-output-rename": true,
    "pipe-naming": [true, "camelCase", "pre"],
    "template-to-ng-template": true,
    "templates-use-public": true,
    "use-host-property-decorator": true,
    "use-input-property-decorator": true,
    "use-life-cycle-interface": true,
    "use-output-property-decorator": true,
    "use-pipe-transform-interface": true,

    // TSLint Overrides/Custom
    "align": [false],
    "interface-name": [false],
    "linebreak-style": [true, "LF"],
    "no-angle-bracket-type-assertion": false,
    "no-empty-interface": false,
    "no-for-in-array": true,
    "no-invalid-template-strings": false,
    "no-namespace": false,
    "no-parameter-properties": false,
    "no-sparse-arrays": true,
    "no-string-literal": false,
    "no-unnecessary-callback-wrapper": true,
    // "no-unused-variable": [true],
    "object-literal-sort-keys": false,
    "ordered-imports": [false],
    "quotemark": [
      true,
      "single",
      "avoid-escape"
    ],
    "prefer-const": false,
    "prefer-template": true,
    "return-undefined": true,
    "trailing-comma": [true, { "multiline": "never", "singleline": "never" }],
    "variable-name": [
      true,
      "allow-leading-underscore",
      "allow-pascal-case",
      "ban-keywords",
      "check-format"
    ]
  }
}

```

#### Actual behavior

ERROR: (no-shadowed-variable) form.component.ts[18, 12]: Shadowed name: 'ValidatorFn'

#### Expected behavior

I would expect ValidatorFn to be allowed here because it's a type assertion.

**Comment by ajafff** [comment]
That's no type assertion, that's a type parameter. So the error is expected.

```ts
let foo = <Foo>() => {};
// equals
let foo = function <Foo>() {};
```

**Comment by jgornick** [comment]
Thanks for the clarification @ajafff!
